src/xdg-autostart-generator/xdg-autostart-generator.c

```diff
@@ -0,0 +1,771 @@
+        escaped = unit_name_escape(c);
+        if (!escaped)
+                return NULL;
+
+        asprintf(&res, "apps-%s-autostart.service", escaped);
+
+        return res;
+}
+
+static int xdg_config_parse_bool(
+                const char *unit,
+                const char *filename,
+                unsigned line,
+                const char *section,
+                unsigned section_line,
+                const char *lvalue,
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata) {
+
+        bool *b = data;
+        const char *value;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(data);
+
+        value = skip_leading_chars(rvalue, NULL);
+        if (streq(value, "true")) {
+                *b = 1;
+        } else if (streq(value, "false")) {
+                *b = 0;
+        } else {
+                log_syntax(unit, LOG_ERR, filename, line, 0, "Invalid value for boolean: %s", value);
+                return 0;
+        }
+
+        return 0;
+}
+
+/* Unescapes the string in-place, returns non-zero status on error. */
+static int xdg_unescape_string(
+                const char *unit,
+                const char *filename,
+                int line,
+                char *str) {
+
+        char *in;
+        char *out;
+
+        assert(str);
+
+        in = out = str;
+
+        for (; *in; in++, out++) {
+                if (*in == '\\') {
+                        /* Move forward, and ensure it is a valid escape. */
+                        in++;
+
+                        switch (*in) {
+                                case 's':
+                                        *out = ' ';
+                                        break;
+                                case 'n':
+                                        *out = '\n';
+                                        break;
+                                case 't':
+                                        *out = '\t';
+                                        break;
+                                case 'r':
+                                        *out = '\r';
+                                        break;
+                                case '\\':
+                                        *out = '\\';
+                                        break;
+                                case ';':
+                                        /* Technically only permitted for strv. */
+                                        *out = ';';
+                                        break;
+                                default:
+                                        log_syntax(unit, LOG_ERR, filename, line, 0, "Undefined escape sequence \\%c.", *in);
+                                        return -EINVAL;
+                        }
+
+                        continue;
+                }
+
+                *out = *in;
+        }
+        *out = '\0';
+
+        return 0;
+}
```

Comment:
This code looks OK, but I'd prefer to have it in a separate file, so that we can add a test for it.
Edit: in fact, after reading to the end, I'm pretty sure we'll want to hook this up to a fuzzer.

Reply:
I placed most of it into xdg-autostart-service.c and tried to hook up some basic testing. Not perfect yet probably, but it hopefully is a good start.

Not sure how to hook it up to a fuzzer.